net/multi_listen.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package net
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"sync/atomic"
+)
+
+// connAndErrPair packs results of accept call on actual listeners together along with the
+// index of the listener. It is used for communication between main and listener goroutines.
+type connAndErrPair struct {
+	index int
+	conn  net.Conn
+	err   error
+}
+
+// multiListener implements net.Listener
+type multiListener struct {
+	latestAcceptedIndex atomic.Int32
+	addrs               []net.Addr
+	listeners           []net.Listener
+
+	wg           sync.WaitGroup
+	connAndErrCh chan connAndErrPair
+	stopCh       chan struct{}
+}
+
+// compile time check to ensure *multiListener implements net.Listener.
+var _ net.Listener = &multiListener{}
+
+// MultiListen returns net.Listener which can listen for and accept
+// TCP connections on multiple addresses.
+func MultiListen(addresses []string) (net.Listener, error) {
+	return multiListen(addresses, net.ResolveTCPAddr, net.Listen)
+}
+
+func multiListen(
+	addresses []string,
+	resolveTCPAddrFunc func(network, address string) (*net.TCPAddr, error),
+	listenFunc func(network, address string) (net.Listener, error),
+) (net.Listener, error) {
+	ml := &multiListener{
+		connAndErrCh: make(chan connAndErrPair),
+		stopCh:       make(chan struct{}),
+	}
+
+	for _, address := range addresses {
```

Comment:
we may need to filter duplicate addresses? or do we want to fail (I think it will fail to create a listener in the same address twice)?

Reply:
I think the caller be responsible for that (not passing the same address twice).
I can't imagine a case where there is a need to listen on the same address twice, if it is passed it might be because of some bug and we should error out instead of silently ignoring the dup address, right?

Reply:
fair

Reply:
In the same area: return an error if the address slice is empty. I think it works like this now: (ml, nil) ia returned, and a subsequent Accept() hangs forever. While this is logical (sort of), it's probably not what the caller intended.